src/lib.rs

```diff
@@ -475,10 +484,210 @@ fn parse_and_test_file(
+                    first_rev.1.messages.get(line + 1).map_or(&[], |m| &**m);
+
+                print_msgs.clear();
+                print_msgs.extend(
+                    first_msgs
+                        .iter()
+                        .filter(|m| m.level == Level::Error)
+                        .filter_map(|m| {
+                            m.line_col
+                                .as_ref()
+                                .zip(m.code.as_deref().and_then(|c| c.strip_prefix(prefix)))
+                        })
+                        .inspect(|(span, _)| use_over_matcher |= span.line_start != span.line_end)
+                        .enumerate()
+                        .map(|(i, (span, code))| (i, span, code, first_rev.0)),
+                );
+                counters.clear();
+                counters.resize(print_msgs.len(), 0);
+
+                for rev in revs {
+                    let msgs: &[Message] = rev.1.messages.get(line + 1).map_or(&[], |m| &**m);
+
+                    for (span, code) in
+                        msgs.iter()
+                            .filter(|m| m.level == Level::Error)
+                            .filter_map(|m| {
+                                m.line_col
+                                    .as_ref()
+                                    .zip(m.code.as_deref().and_then(|c| c.strip_prefix(prefix)))
+                            })
+                    {
+                        let i = if let Some(&(i, ..)) = print_msgs[..counters.len()].iter().find(
+                            |&&(_, prev_span, prev_code, _)| span == prev_span && code == prev_code,
+                        ) {
+                            counters[i] += 1;
+                            i
+                        } else {
+                            use_over_matcher |= span.line_start != span.line_end;
+                            usize::MAX
+                        };
+                        print_msgs.push((i, span, code, rev.0));
+                    }
+                }
+
+                // partition the first revision's messages
+                // in all revisions => only some revisions
+                let mut i = 0;
+                let mut j = counters.len();
+                while i < j {
+                    if counters[i] == revs.len() {
+                        print_msgs[i].3 = "";
+                        i += 1;
+                    } else if counters[j - 1] == revs.len() {
+                        print_msgs.swap(i, j - 1);
+                        print_msgs[i].3 = "";
+                        i += 1;
+                        j -= 1;
+                    } else {
+                        j -= 1;
+                    }
+                }
+                // For all other revision's messages, remove the ones that exist in all revisions.
+                print_msgs.retain(|&(i, _, _, rev)| {
+                    rev.is_empty() || counters.get(i).map_or(true, |&x| x != revs.len())
+                });
```

Comment:
maybe do a filter pass on the full data first, and then just print out everything at once instead of mixing the printing and the filtering

Reply:
I don't see how this comment matches the lines you're commenting on. Nothing has been printed by this point.

Reply:
This filtering out of duplicated messages could be a separate pass before the printing. Here it is in the middle of printing logic

Reply:
The printing code follows immediately after this, it's not in the middle.